tests/raft_log_replication.py

```diff
@@ -593,7 +593,441 @@ async def _unittest_raft_log_replication() -> None:
+    assert raft_node_2._log[0].entry.value == 0
+    assert raft_node_2._log[1].term == 4
+    assert raft_node_2._log[1].entry.name.value.tobytes().decode("utf-8") == "top_1"
+    assert raft_node_2._log[1].entry.value == 7
+    assert raft_node_2._log[2].term == 5
+    assert raft_node_2._log[2].entry.name.value.tobytes().decode("utf-8") == "top_2"
+    assert raft_node_2._log[2].entry.value == 8
+    assert raft_node_2._log[3].term == 6
+    assert raft_node_2._log[3].entry.name.value.tobytes().decode("utf-8") == "top_3"
+    assert raft_node_2._log[3].entry.value == 9
+    assert raft_node_2._commit_index == 3
+
+    assert len(raft_node_3._log) == 1 + 3
+    assert raft_node_3._log[0].term == 0
+    assert raft_node_3._log[0].entry.name.value.tobytes().decode("utf-8") == ""  # index zero entry is empty
+    assert raft_node_3._log[0].entry.value == 0
+    assert raft_node_3._log[1].term == 4
+    assert raft_node_3._log[1].entry.name.value.tobytes().decode("utf-8") == "top_1"
+    assert raft_node_3._log[1].entry.value == 7
+    assert raft_node_3._log[2].term == 5
+    assert raft_node_3._log[2].entry.name.value.tobytes().decode("utf-8") == "top_2"
+    assert raft_node_3._log[2].entry.value == 8
+    assert raft_node_3._log[3].term == 6
+    assert raft_node_3._log[3].entry.name.value.tobytes().decode("utf-8") == "top_3"
+    assert raft_node_3._log[3].entry.value == 9
+    assert raft_node_3._commit_index == 3
+
+    assert len(raft_node_4._log) == 1 + 3
+    assert raft_node_4._log[0].term == 0
+    assert raft_node_4._log[0].entry.name.value.tobytes().decode("utf-8") == ""  # index zero entry is empty
+    assert raft_node_4._log[0].entry.value == 0
+    assert raft_node_4._log[1].term == 4
+    assert raft_node_4._log[1].entry.name.value.tobytes().decode("utf-8") == "top_1"
+    assert raft_node_4._log[1].entry.value == 7
+    assert raft_node_4._log[2].term == 5
+    assert raft_node_4._log[2].entry.name.value.tobytes().decode("utf-8") == "top_2"
+    assert raft_node_4._log[2].entry.value == 8
+    assert raft_node_4._log[3].term == 6
+    assert raft_node_4._log[3].entry.name.value.tobytes().decode("utf-8") == "top_3"
+    assert raft_node_4._log[3].entry.value == 9
+    assert raft_node_4._commit_index == 3
+    
+    await asyncio.sleep(TERM_TIMEOUT + 1)
+    
+    _logger.info("================== TEST 2: Leadership Change to Node 42 and Add New Entry  ==================")
+
+    # New LEADER => raft_node_2
+    raft_node_1.election_timeout = ELECTION_TIMEOUT
+    raft_node_2.election_timeout = ELECTION_TIMEOUT + 1
+    raft_node_3.election_timeout = ELECTION_TIMEOUT + 2
+    raft_node_4.election_timeout = ELECTION_TIMEOUT + 2.1
+    await asyncio.sleep(TERM_TIMEOUT)
+
+    raft_node_1.close() # Simulation of the disappearance of a leader from a cluster
+
+    await asyncio.sleep(ELECTION_TIMEOUT + 8*TERM_TIMEOUT + 0.1) # Six terms are needed for complete replication of logs from the leader to the followers.
```

Comment:
Ok, I think I understand your issue here.

Currently when a node becomes a new leader it assumes all the other nodes are behind/have empty logs.

If you think about it, this only makes sense at the start, after that the nodes will always be in various degrees of up-to-dateness.

This means that when a node becomes Leader, the best strategy is to assume all other nodes _are_ up-to-date, until proven otherwise. After sending the first Heartbeat/AppendEntryRequest, you will now whether that is true:

<img width="443" alt="Screenshot 2024-07-10 at 16 42 51" src="https://github.com/OpenCyphal-Garage/cyraft/assets/39975120/28053e02-7f8a-4177-b4b8-4108abc5682f">

Then it should start to send AppendEntriesRequest in decreasing order, I think this part is already implemented:

<img width="775" alt="image" src="https://github.com/OpenCyphal-Garage/cyraft/assets/39975120/a70a59c4-0db5-4e1f-a61d-570747891bd2">

Now I'm also realising we're only testing the case where all the other nodes are indeed up-to-date. So therefore it would make sense to, once `node_2` has been elected as leader, to clear `node_1`'s log and re-start it, this node should then also be brought up to date by `node_2`.